### ISSUE TITLE
Index rebuilds should refuse to run in parallel

### ISSUE THREAD

**Issue by gregorydlogan** [opencast/opencast#1451]
As described [here](https://groups.google.com/a/opencast.org/forum/m/#!topic/dev/3wiAfLGlPfc), the elastic search indexes do not properly reindex if the re indexing is run in parallel. This seems like it would be easily solved with a semaphore and some OSGI glue.

**Comment by lkiesow** [comment]
No, you actually need to rewrite the whole index code since the reindexing is already started sequentially and not in parallel. But re-indexing works by sending messages via network. A process that does not necessarily guarantee the order of the messages. You probably heard me complain about this in the past already ;-)

**Comment by gregorydlogan** [comment]
I'm not terribly familiar with the code in question, but if the issue is that the results of index B depend on index A we could at least solve some of the issue by enforcing that only a single index be rebuilt at a time, right?  It doesn't guarantee that index A hasn't been rebuilt with bad data, but at least it prevents the user from kicking off all of the indexing at once.

**Comment by lkiesow** [comment]
We are already doing that. In a way…
We are building index A and sending all the necessary messages. Only if that's done we start with index B. Unfortunately, sending all messages does not mean that all messages are already received and processed. That code is completely decoupled and isn't even directly related to index-rebuilds at all.

What we usually do for updates is that we set the log level for the message receiver to debug and wait until all messages are seemingly processed before initiating the rebuilding process for the next index. But that's a bit hacky.

We want to fix that for SWITCH and already started the development. But then CORVID19 happened and we need to scale up and do weird other things :D

**Comment by gregorydlogan** [comment]
Fair enough.  I'll assign this to you then, since you're already working on it, and SWITCH will make you finish it :)

**Comment by lkiesow** [comment]
Fair enough. I'll assign this to @KatrinIhler then, since she is already working on it, and SWITCH will make her finish it :)

**Comment by karendolan** [comment]
@KatrinIhler , was this fixed in a pull?

**Comment by KatrinIhler** [comment]
@karendolan This is currently in progress! If you look at my open pull requests right now, starting at #2211, I'm working on rewriting how the index rebuild works. Part of this is ripping out the messaging stuff, so once this is done, this issue should be fixed. :D If you want to see the change for one service, check out #2351 for example - the other services should work similarly. 3 are already done, with 5 more to go! (Unfortunately, I started with the easy ones...)

**Comment by lkiesow** [comment]
This work was dome by @KatrinIhler and is in 10.1 as explained in #2893

**Comment by KatrinIhler** [comment]
Purely for completeness' sake: You can still do an index rebuild in parallel if you make multiple requests to the index rebuild service. There is currently no check forbidding that. (And maybe there should be, because this is a really bad idea in most - but not all - cases.) But if you start a complete index rebuild _once_, everything happening then is now guaranteed to be sequential.
